### PR TITLE
Special handling for users endpoint in auth adapter (GSI-645)

### PR DESCRIPTION
**Pull request by Cito**
This PR implements the two endpoints for creating and updating users in the auth adapter, since the need special treatment. In these cases, an internal auth token needs to be created even though the user is not yet fully authenticated.

The PR also adds the status and role fields to the user session.

### CODE REVIEW

File: src/auth_service/auth_adapter/api/main.py [hunk 1]
```diff
@@ -189,6 +208,68 @@ async def logout(
     return Response(status_code=204)
 
 
+@app.post(
+    "/users",
+    operation_id="post_user",
+    tags=["users"],
+    summary="Register a user",
+    description="Authenticate the endpoint used to register a new user.",
```
Comment (mephenor) [inline src/auth_service/auth_adapter/api/main.py:216]:
```suggestion
    description="Authenticate against the endpoint used to register a new user.",
```

File: src/auth_service/auth_adapter/api/main.py [hunk 2]
```diff
@@ -189,6 +208,68 @@ async def logout(
...
+    response.headers["Authorization"] = authorization
+    return response
+
+
+@app.put(
+    "/users/{id}",
+    operation_id="post_user",
+    tags=["users"],
+    summary="Update a user",
+    description="Authenticate the endpoint used to update an existing user.",
```
Comment (mephenor) [inline src/auth_service/auth_adapter/api/main.py:241]:
```suggestion
    description="Authenticate against the endpoint used to update an existing user.",
```

File: src/auth_service/auth_adapter/api/main.py [hunk 3]
```diff
@@ -189,6 +208,68 @@ async def logout(
     return Response(status_code=204)
 
 
+@app.post(
+    "/users",
+    operation_id="post_user",
+    tags=["users"],
+    summary="Register a user",
```
Comment (mephenor) [inline src/auth_service/auth_adapter/api/main.py:215]: Not sure if I'm missing something, but the summary and docstring are a bit confusing. What does register mean in this context? Is it implicit or should the code also set the session state to registered and that line of code is missing?

File: src/auth_service/auth_adapter/api/main.py [hunk 4]
```diff
@@ -189,6 +208,68 @@ async def logout(
...
+    authorization = f"Bearer {internal_token}"
+    response.headers["Authorization"] = authorization
+    return response
+
+
+@app.put(
+    "/users/{id}",
+    operation_id="post_user",
+    tags=["users"],
+    summary="Update a user",
```
Comment (mephenor) [inline src/auth_service/auth_adapter/api/main.py:240]: Same problem as with the other endpoint.

File: tests/integration/auth_adapter/test_sessions.py [hunk 5]
```diff
@@ -267,6 +272,21 @@ async def test_login_with_registered_user_with_title(
     )
 
 
+@mark.asyncio
+async def test_login_with_deactivated_user(
+    client: AsyncTestClient, httpx_mock: HTTPXMock
+):
+    """Test that a login request can create a new session for a registered user."""
```
Comment (mephenor) [inline tests/integration/auth_adapter/test_sessions.py:279]:
```suggestion
    """Test that a login request cannot create a new session for an inactive user."""
```

**Review by mephenor — CHANGES_REQUESTED** [review]
As always mostly some minor stuff.
There's also minor clarification and perhaps some minor changes needed for both of them.

File: src/auth_service/auth_adapter/api/main.py [hunk 6]
```diff
@@ -189,6 +208,68 @@ async def logout(
...
+) -> Response:
+    """Register a user."""
+    response = Response(status_code=200)
+    if not session:
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED, detail="Not logged in"
+        )
+    check_csrf("POST", x_csrf_token, session)
+    internal_token = internal_token_from_session(session)
+    authorization = f"Bearer {internal_token}"
```
[the remark this answers is not in the document]
Comment (Cito) [reply src/auth_service/auth_adapter/api/main.py:231]: Yes, I thought the same and had it already fixed in the upcoming PR that is based on this one. So will leave it like that in this PR.

File: src/auth_service/auth_adapter/api/main.py (hunk 3 above)
Comment (Cito) [reply src/auth_service/auth_adapter/api/main.py:215]: Yes, it's a bit confusing. The endpoint here doesn't register anything. It just allows (by returning status 200) to use the endpoint with the same path in the user registry. And that endpoint then registers the user. I.e. saves the user to the database that is stored in the session and passed via internal auth token.

Comment (Cito) [reply src/auth_service/auth_adapter/api/main.py:215]: Will be improved in the next PR.

File: tests/integration/auth_adapter/test_sessions.py (hunk 5 above)
Comment (Cito) [reply tests/integration/auth_adapter/test_sessions.py:279]: Thanks, will be fixed in upcoming PR.

File: src/auth_service/auth_adapter/api/main.py (hunk 2 above)
Comment (Cito) [reply src/auth_service/auth_adapter/api/main.py:241]: Actually, I meant to authenticate the endpoint itself (i.e. allowing that the endpoint can be used). Will try to improve the docstring in the next PR.

**Review by mephenor — APPROVED** [review]

**Pull request merged**